src/nvim/getchar.c

```diff
@@ -2312,7 +2317,7 @@ static int vgetorpeek(const int advance)
     }           /* if (!character from stuffbuf) */
 
     /* if advance is FALSE don't loop on NULs */
-  } while (c < 0 || (advance && c == NUL));
+  } while (!exiting);
```

Comment:
Some of the commits I have here are essentially "showing my workings", where I've split up the change I want to make into some changes that are much easier to verify are correct.

This is one of the commits in the series that splits the multiple uses of the variable `c` into separate variables.
Initially it was used to store each of

- The final character to return
- The number of bytes read from the user
- A `-1` put there simply to say "continue on in the loop"

at various points in the `while()` loop.

I found this very confusing, which is why I split it up.
The `exiting` variable is a temporary one (removed in later commits), that removes one of the uses, so that the others are easy to reason about for future changes.